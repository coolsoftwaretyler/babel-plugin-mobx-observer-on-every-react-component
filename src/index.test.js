import { expect, test, describe } from "bun:test";
import { transform } from "@babel/core";
import pluginSyntaxJsx from "@babel/plugin-syntax-jsx";
import pluginTransReactJsx from "@babel/plugin-transform-react-jsx";
import { autoObserverPlugin } from './index.mjs'

const runTransform = (input) => {
  return transform(input, {
    plugins: [pluginSyntaxJsx, pluginTransReactJsx, autoObserverPlugin]
  });
};

describe("Mobx Observer Babel Plugin", () => {
  describe('when there is no mobx-react import', () => {
    describe('for a single arrow function component', () => {
      test('it imports observer and wraps the component', () => {
        const testinput = `const Simple = () => <div>check</div>;`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('for multiple arrow function components', () => {
      test('it imports observer and wraps both components', () => {
        const testinput = `const Simple = () => <div>check</div>;
        const Simple2 = () => <div>check</div>;`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('for a single functional declaration', () => {
      test('it imports observer and wraps the component', () => {
        const testinput = `function Simple() { return <div>check</div>; }`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('for multiple functional declarations', () => {
      test('it imports observer and wraps both components', () => {
        const testinput = `function Simple() { return <div>check</div>; }
        function Simple2() { return <div>check</div>; }`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('for a single function expression', () => {
      test('it imports observer and wraps the component', () => {
        const testinput = `const Simple = function() { return <div>check</div>; }`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('for multiple function expressions', () => {
      test('it imports observer and wraps both components', () => {
        const testinput = `const Simple = function() { return <div>check</div>; }
        const Simple2 = function() { return <div>check</div>; }`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('for a single class declaration', () => {
      test('it imports observer and wraps the component', () => {
        const testinput = `class Simple extends React.Component { render() { return <div>check</div>; } }`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('for multiple class declarations', () => {
      test('it imports observer and wraps both components', () => {
          const testinput = `class Simple extends React.Component { render() { return <div>check</div>; } }
        class Simple2 extends React.Component { render() { return <div>check</div>; } }`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('when there are additional imports', () => {
      test('it leaves the other imports alone', () => {
        const testinput = `import { types } from "mobx-state-tree";
        const Simple = () => <div>check</div>;`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('when there is no react component at all', () => {
        test('it leaves the code unchanged', () => {
            const testinput = `const Simple = (a, b) => a + b
            const obj = {
                a: 1,
                b: 2,
                c: 3
            }
            `
            const out = runTransform(testinput);

            expect(out.code).toMatchSnapshot();
        })
    })
  })
  describe('when there is a mobx-react import', () => {
    describe('for a single arrow function component', () => {
      test('it leaves the import alone and wraps the component', () => {
        const testinput = `import { observer } from "mobx-react";
        const Simple = () => <div>check</div>;`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })      
    })

    describe('for multiple arrow function components', () => {
      test('it imports observer and wraps both components', () => {
        const testinput = `import { observer } from "mobx-react";
        const Simple = () => <div>check</div>;
        const Simple2 = () => <div>check</div>;`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('for a single functional declaration', () => {
      test('it imports observer and wraps the component', () => {
        const testinput = `import { observer } from "mobx-react";
        function Simple() { return <div>check</div>; }`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('for multiple functional declarations', () => {
      test('it imports observer and wraps both components', () => {
        const testinput = `import { observer } from "mobx-react";
        function Simple() { return <div>check</div>; }
        function Simple2() { return <div>check</div>; }`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('for a single function expression', () => {
      test('it imports observer and wraps the component', () => {
        const testinput = `import { observer } from "mobx-react";
        const Simple = function() { return <div>check</div>; }`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('for multiple function expressions', () => {
      test('it imports observer and wraps both components', () => {
        const testinput = `import { observer } from "mobx-react";
        const Simple = function() { return <div>check</div>; }
        const Simple2 = function() { return <div>check</div>; }`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('for a single class declaration', () => {
      test('it imports observer and wraps the component', () => {
        const testinput = `import { observer } from "mobx-react";
        class Simple extends React.Component { render() { return <div>check</div>; } }`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('for multiple class declarations', () => {
      test('it imports observer and wraps both components', () => {
        const testinput = `import { observer } from "mobx-react";
        class Simple extends React.Component { render() { return <div>check</div>; } }
        class Simple2 extends React.Component { render() { return <div>check</div>; } }`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })

    describe('when there are additional imports', () => {
      test('it leaves the other imports alone', () => {
        const testinput = `import { types } from "mobx-state-tree";
        import { observer } from "mobx-react";
        const Simple = () => <div>check</div>;`
        const out = runTransform(testinput);

        expect(out.code).toMatchSnapshot();
      })
    })
  })
});