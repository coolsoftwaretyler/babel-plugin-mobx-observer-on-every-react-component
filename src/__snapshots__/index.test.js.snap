// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Mobx Observer Babel Plugin when there is no mobx-react import for a single arrow function component it imports observer and wraps the component 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(() => <div>check</div>);"
`;

exports[`Mobx Observer Babel Plugin when there is no mobx-react import for multiple arrow function components it imports observer and wraps both components 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(() => <div>check</div>);
const Simple2 = observer(() => <div>check</div>);"
`;

exports[`Mobx Observer Babel Plugin when there is no mobx-react import for a single functional declaration it imports observer and wraps the component 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(function Simple() {
  return <div>check</div>;
});"
`;

exports[`Mobx Observer Babel Plugin when there is no mobx-react import for multiple functional declarations it imports observer and wraps both components 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(function Simple() {
  return <div>check</div>;
});
const Simple2 = observer(function Simple2() {
  return <div>check</div>;
});"
`;

exports[`Mobx Observer Babel Plugin when there is no mobx-react import for a single function expression it imports observer and wraps the component 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(function () {
  return <div>check</div>;
});"
`;

exports[`Mobx Observer Babel Plugin when there is no mobx-react import for multiple function expressions it imports observer and wraps both components 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(function () {
  return <div>check</div>;
});
const Simple2 = observer(function () {
  return <div>check</div>;
});"
`;

exports[`Mobx Observer Babel Plugin when there is no mobx-react import for a single class declaration it imports observer and wraps the component 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(class Simple extends React.Component {
  render() {
    return <div>check</div>;
  }
});"
`;

exports[`Mobx Observer Babel Plugin when there is no mobx-react import for multiple class declarations it imports observer and wraps both components 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(class Simple extends React.Component {
  render() {
    return <div>check</div>;
  }
});
const Simple2 = observer(class Simple2 extends React.Component {
  render() {
    return <div>check</div>;
  }
});"
`;

exports[`Mobx Observer Babel Plugin when there is no mobx-react import when there are additional imports it leaves the other imports alone 1`] = `
"import { observer } from "mobx-react";
import { types } from "mobx-state-tree";
const Simple = observer(() => /*#__PURE__*/React.createElement("div", null, "check"));"
`;

exports[`Mobx Observer Babel Plugin when there is a mobx-react import for a single arrow function component it leaves the import alone and wraps the component 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(() => /*#__PURE__*/React.createElement("div", null, "check"));"
`;

exports[`Mobx Observer Babel Plugin when there is a mobx-react import for multiple arrow function components it imports observer and wraps both components 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(() => /*#__PURE__*/React.createElement("div", null, "check"));
const Simple2 = observer(() => /*#__PURE__*/React.createElement("div", null, "check"));"
`;

exports[`Mobx Observer Babel Plugin when there is a mobx-react import for a single functional declaration it imports observer and wraps the component 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(function Simple() {
  return <div>check</div>;
});"
`;

exports[`Mobx Observer Babel Plugin when there is a mobx-react import for multiple functional declarations it imports observer and wraps both components 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(function Simple() {
  return <div>check</div>;
});
const Simple2 = observer(function Simple2() {
  return <div>check</div>;
});"
`;

exports[`Mobx Observer Babel Plugin when there is a mobx-react import for a single function expression it imports observer and wraps the component 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(function () {
  return <div>check</div>;
});"
`;

exports[`Mobx Observer Babel Plugin when there is a mobx-react import for multiple function expressions it imports observer and wraps both components 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(function () {
  return <div>check</div>;
});
const Simple2 = observer(function () {
  return <div>check</div>;
});"
`;

exports[`Mobx Observer Babel Plugin when there is a mobx-react import for a single class declaration it imports observer and wraps the component 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(class Simple extends React.Component {
  render() {
    return <div>check</div>;
  }
});"
`;

exports[`Mobx Observer Babel Plugin when there is a mobx-react import for multiple class declarations it imports observer and wraps both components 1`] = `
"import { observer } from "mobx-react";
const Simple = observer(class Simple extends React.Component {
  render() {
    return <div>check</div>;
  }
});
const Simple2 = observer(class Simple2 extends React.Component {
  render() {
    return <div>check</div>;
  }
});"
`;

exports[`Mobx Observer Babel Plugin when there is a mobx-react import when there are additional imports it leaves the other imports alone 1`] = `
"import { types } from "mobx-state-tree";
import { observer } from "mobx-react";
const Simple = observer(() => /*#__PURE__*/React.createElement("div", null, "check"));"
`;
